### PR TITLE
feat(presets): add support for local presets

### LINKED ISSUE #3019: Allow self-hosted presets

### PR DESCRIPTION
**Pull request by fgreinacher**
Closes #3019

🛠  with :heart: by [Siemens](https://github.com/siemens)

### CODE REVIEW

File: docs/usage/config-presets.md [hunk 1]
```diff
@@ -168,6 +168,10 @@ To host your preset config on GitLab:
 
 Note: Unlike npmjs-hosted presets, GitLab-hosted ones can contain only one config.
 
+## Local presets
+
+Renovate also supports local presets, i.e. presets that are hosted on the same platform as the target repository. This is especially helpful in self-hosted scenarios where public presets cannot be used. Local presets are only supported on GitHub and GitLab. Local presets are referenced via the `local>` prefix, e.g. `local>owner/name`. Renovate will determine the current platform and look up the preset from there.
+
```
Comment (fgreinacher) [inline docs/usage/config-presets.md]: Happy to get some feedback on this!

File: lib/config/presets/gitlab.ts [hunk 2]
```diff
@@ -4,12 +4,11 @@ import { Preset } from './common';
 
 const { get: glGot } = api;
```
Comment (fgreinacher) [inline lib/config/presets/gitlab.ts]: Will the token from the config automatically be used for API calls?

**Comment by fgreinacher** [comment]
This is now ready for review.

File: lib/config/presets/gitlab.ts (hunk 2 above)
Comment (viceice) [reply lib/config/presets/gitlab.ts]: Yes, i think so.

File: lib/config/common.ts [hunk 3]
```diff
@@ -121,6 +121,7 @@ export interface RenovateConfig
   branchList?: string[];
   description?: string[];
 
+  platform?: string;
```
Comment (fgreinacher) [inline lib/config/common.ts]: Was missing - I assume just because it was not used after TS migration?

**Comment by fgreinacher** [comment]
macOS test failure seems unrelated to my changes

File: lib/config/presets/gitlab.ts [hunk 4]
```diff
@@ -4,12 +4,11 @@ import { Preset } from './common';
...
 
-const GitLabApiUrl = 'https://gitlab.com/api/v4/projects';
-
 async function getDefaultBranchName(
-  urlEncodedPkgName: string
+  urlEncodedPkgName: string,
+  endpoint: string
 ): Promise<string> {
-  const branchesUrl = `${GitLabApiUrl}/${urlEncodedPkgName}/repository/branches`;
+  const branchesUrl = `${endpoint}/projects/${urlEncodedPkgName}/repository/branches`;
```
Comment (rarkins) [inline lib/config/presets/gitlab.ts]: Does this work ok if `endpoint` is defined with a trailing `/` ?

Comment (fgreinacher) [reply lib/config/presets/gitlab.ts]: Nope, thanks! Fixed with dbb0debbacff04c9ad7bf1508aa02e151fecf567

**Comment by fgreinacher** [comment]
Adding back WIP until I had the chance to test "for real" this on our GitLab instance, early next week.

**Comment by fgreinacher** [comment]
Verified this and found that I did not properly pass the current/base config to the preset resolution logic. This is fixed with f2854801b67d4126dbce05863678a7584706b548.

File: lib/config/presets/local.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import * as github from './github';
+import { RenovateConfig } from '../common';
+
+export async function getPreset(
+  pkgName: string,
+  presetName = 'default',
+  baseConfig: RenovateConfig
+): Promise<Preset> {
+  if (baseConfig.platform?.toLowerCase() === 'gitlab') {
+    return gitlab.getPreset(pkgName, presetName, baseConfig.endpoint);
```
Comment (fgreinacher) [inline lib/config/presets/local.ts]: I am unsure whether it is good to have the switch here vs. moving this logic (or parts of it) to the `platform` module. What do you think?

Comment (rarkins) [reply lib/config/presets/local.ts]: Let's keep a switch here for now. We support 6 platforms now but only 2 of them with presets, so I think it's ok to keep it in the presets logic.

Comment (fgreinacher) [reply lib/config/presets/local.ts]: Alright!

File: lib/config/common.ts (hunk 3 above)
Comment (viceice) [inline lib/config/common.ts]: I think this should be moved to `RenovateAdminConfig` interface

**Review by viceice — CHANGES_REQUESTED** [review]
Would like to see it for bitbucket server too 😅

File: docs/usage/config-presets.md [hunk 6]
```diff
@@ -168,6 +168,10 @@ To host your preset config on GitLab:
 
 Note: Unlike npmjs-hosted presets, GitLab-hosted ones can contain only one config.
 
+## Local presets
+
+Renovate also supports local presets, i.e. presets that are hosted on the same platform as the target repository. This is especially helpful in self-hosted scenarios where public presets cannot be used. Local presets are only supported on GitHub and GitLab. Local presets are referenced via the `local>` prefix, e.g. `local>owner/name`. Renovate will determine the current platform and look up the preset from there.
```
Comment (viceice) [inline docs/usage/config-presets.md]: @rarkins  Why not just use `>` ? eg `>owner/name`. 🤔

Comment (rarkins) [reply docs/usage/config-presets.md]: Actually this triggered my memory that we could perhaps use *no* prefix as discussed in #3019. @fgreinacher what do you think about that? According to someone who tested it, a preset `a/b` today will attempt to look up npm package `a%2fb` which can never exist. i.e. `a/b` is an "available" and perhaps intuitive namespace to use? i.e. "if the preset name contains one or more `/` and does not start with `@`, then assume it is a local repository"

Comment (fgreinacher) [reply docs/usage/config-presets.md]: Not a strong opinion, but I'd prefer being explicit here

Comment (viceice) [reply docs/usage/config-presets.md]: i prefer shorter configs 😅

**Comment by fgreinacher** [comment]
> Would like to see it for bitbucket server too 😅

Go for it 🥇

File: lib/config/common.ts (hunk 3 above)
Comment (fgreinacher) [reply lib/config/common.ts]: `endpoint` also?

Comment (viceice) [reply lib/config/common.ts]: Yes, i think so.

File: docs/usage/config-presets.md (hunk 6 above)
Comment (fgreinacher) [reply docs/usage/config-presets.md]: @rarkins Your decision 👨‍✈

File: lib/config/common.ts (hunk 3 above)
Comment (fgreinacher) [reply lib/config/common.ts]: ✅ 8e94ce70c5da86c752adece72058849f7e72efb0

**Comment by fgreinacher** [comment]
[Windows failure](https://github.com/renovatebot/renovate/pull/5814/checks?check_run_id=549132914) seems to be unrelated

**Comment by viceice** [comment]
Restarted job

File: docs/usage/config-presets.md (hunk 6 above)
Comment (rarkins) [reply docs/usage/config-presets.md:173]: @fgreinacher in that case let's do both :)

`a/b` and `local>a/b` being equivalent. Can you update this PR?

Comment (fgreinacher) [reply docs/usage/config-presets.md:173]: Sure, will do :)

File: lib/config/presets/index.ts [hunk 7]
```diff
@@ -66,6 +66,12 @@ export function parsePreset(input: string): ParsedPreset {
   } else if (str.startsWith('local>')) {
     presetSource = 'local';
     str = str.substring('local>'.length);
+  } else if (
+    !str.startsWith('@') &&
+    !str.startsWith(':') &&
```
Comment (fgreinacher) [inline lib/config/presets/index.ts:71]: To prevent conflicts with built-in presets.

File: docs/usage/config-presets.md (hunk 6 above)
Comment (fgreinacher) [reply docs/usage/config-presets.md]: Done with fb04453728bc9fbbf4e8f9119f36a81217075748

**Comment by fgreinacher** [comment]
Now something seems to be wrong with the macOS jobs 🤖

**Review by viceice — DISMISSED** [review]

**Comment by fgreinacher** [comment]
macOS job seems to need another restart 😐

**Review by viceice — DISMISSED** [review]

**Review by rarkins — CHANGES_REQUESTED** [review]
When I test this locally, it fails due to what seems like a pretty simple bug.

github.getPreset()'s third argument should be the endpoint string: https://github.com/fgreinacher/renovate/blob/bcc9f1972ca2f0e099b5824516440e9dd10f21df/lib/config/presets/github.ts#L48

But here you can see it's called using a RenovateConfig object for endpoint instead of string: https://github.com/fgreinacher/renovate/blob/bcc9f1972ca2f0e099b5824516440e9dd10f21df/lib/config/presets/index.ts#L124

@viceice seems like our typings didn't work in this case - can we address that while we're at it?

**Comment by viceice** [comment]
🤔  This should be a compiler error. Looking into

**Comment by rarkins** [comment]
@viceice could you still check why the types weren't working here, in case the fix reveals any more problems?

**Comment by rarkins** [comment]
I've done a simple test again and it works for me on github.

**Comment by fgreinacher** [comment]
> @viceice could you still check why the types weren't working here, in case the fix reveals any more problems?

Although I'm not a TypeScript expert it seems like the call to `getPreset` is considered untyped:

```ts
let presetConfig = await presetSources[presetSource].getPreset(...);
```

Where `presetSources` is a plain object:

```ts
const presetSources = {
  github,
  npm,
  gitlab,
  local,
};
```

Would be crazy if the TypeScript compiler could understand this usage pattern and "unify" the sources :)

I'd guess something like this would work:

```ts
const presetSources = new Map([ [ "github", github ], /* ... */ ]);

// ...

let presetConfig = await presetSources[presetSource].getPreset(/* ... */);
```

 WDYT?

**Comment by rarkins** [comment]
> WDYT?

I think you're on the right track. We probably need to define an interface for presets first, but it's outside the scope for this PR.

**Pull request merged**